UltraStar Play/Assets/src/model/song/MutableVoice.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+// this should be internal but tests become impossible
+public class MutableVoice
+{
+    private readonly string m_name;
+    private readonly List<MutableSentence> m_sentences = new List<MutableSentence>();
+
+    public MutableVoice(string name)
+    {
+        if (name == null)
+        {
+            throw new ArgumentNullException("name");
+        }
+        m_name = name;
+    }
+
+    public string GetName()
+    {
+        return m_name;
+    }
+
+    public void AddSentence(MutableSentence sentence)
+    {
+        if (sentence == null)
+        {
+            throw new ArgumentNullException("sentence");
+        }
+        else if (sentence.IsEmpty())
+        {
+            Debug.LogWarning("Ignoring empty sentence (hint: it ends at linebreak: " + sentence.GetLinebreakBeat() + ")");
+        }
+        else
+        {
+            sentence.AssertValidNotes();
+            m_sentences.Add(sentence);
+        }
+    }
+
+    public Voice AsVoice()
+    {
+        AssertValidSentences();
+        List<Sentence> sentences = new List<Sentence>();
+        foreach (MutableSentence ms in m_sentences)
+        {
+            sentences.Add(new Sentence(ms.GetNotes(), ms.GetLinebreakBeat()));
+        }
+        return new Voice(m_name, sentences);
+    }
+
+    private void AssertValidSentences()
+    {
+        // assert that there are any sentences at all, do not silently drop a voice
+        if (m_sentences.Count == 0)
+        {
+            throw new SongBuilderException("Voice '" + m_name + "' has no sentences");
+        }
+        // assert sorted sentences
+        m_sentences.Sort((a, b) => a.GetStartBeat().CompareTo(b.GetStartBeat()));
```

Comment:
I wasn't sure if I could assume (read: completely reject a song if it did not confirm to this) that all the sentences (and also the notes within a sentence) should be sequential in the .txt file itself (ie, the beat of the next thing (not counting duets) should always be equal to or larger than the previous one.

If I can make that assumption I will just assert it during the parsing and if it doesn't fail we know it will be sorted.

Reply:
I'd say this assumption is safe to make. I know that there are songs out there that will fail this check. In usdx, these lines would be skipped / passed quickly within 1 frame, so the user would not have been able to hit these txt lines anyways and miss a bunch of points. Better make people fix their faulty song files. (or at least that is my person opinion on this. If you prefer to do the check + sorting and think it is worth it to spend that loading time, go ahead. Just be awesome 😄 )

Reply:
There is now a separate method on the SongMeta object called "GetVoices," which underwater opens the txt and only _then_ actually parses the contents. I also made parsing much simpler; the txt will now just have to be valid (no overlapping notes, no empty voices, no strange linebreaks, etc) otherwise it will just throw an exception.